libs/functions/calculate_projections.py

```diff
@@ -108,46 +119,57 @@ def get_state_projections_df(input_dir, initial_intervention_type, state_interve
     """
     for each state in our data look at the results we generated via run.py
     to create the projections
+
+    columns=CALCULATED_PROJECTION_HEADERS_STATES
     """
 
-    # save results in a list of lists, converted to df later
-    results = []
-    for state in list(US_STATE_ABBREV.values()):
-        intervention_type = _get_intervention_type(initial_intervention_type, state, state_interventions_df)
+    sdf = pd.DataFrame(US_STATE_ABBREV.values(), columns=['state'])
+    sdf.loc[:,'intervention_type'] = sdf.state.parallel_apply(
+        lambda x: _get_intervention_type(initial_intervention_type, x, state_interventions_df)
+        )
+    sdf.loc[:, 'path'] = sdf.parallel_apply(
+        lambda x: get_file_path(input_dir, x.state, x.intervention_type, fips=None),
+        axis=1).values
+    ndf = sdf.parallel_apply(
+        lambda x:_calculate_projection_data(x.state, x.path, fips=None), axis=1)
```

Comment:
I'm a bit hesitant at this refactor - I like that the intervention type and path are separated out from the calculate projection data part (so that we can in fact parallelize those operations), but it seems unnecessary to use pandas.
```python
interventions_and_paths = []
for state in US_STATE_ABBREV.values():
    path = get_path()
    intervention = get_intervention_type()

pool = Pool(10 or w/e)
projection_data = pool.map(_calculate_projection_data, interventions_and_path)
missing = [result for result in projection_data if result is None]
results = [result for result in projection_data if result is not None]
return pd.DataFrame(results, columns=CALCULATED_PROJECTION_HEADERS_STATES)
```

Reply:
just a thought there ^

Reply:
Yeah, I understand.  I started with the same code shown above, but went away form multiprocessing, and even tried joblib, but for a variety of technical, complexity issues, and practices, i think it's best to use `apply` with parallelization.


1. My understanding is that using "for loops" is a not a best practice in pandas.  And removing for loops improved performance when going from iterrows => apply.

2. In addition, with this type of refactor, if something needs parallelization, we can just simply change it from apply to parallel_apply.

3. Wrt parallelizatio nitself, started with multiprocessing.pool.  Ran into issues pickling the data, (just using tuples and lists). As a result, I had to write specific functions that just handled how data was being returned to specific rows.   After that, we have to reaggregate the rows, and handle rows that are null differently from rows that are full.  So now we have mappers that return multiple types -> anti-pattern.

There's just a lot of code to do the basic stuff that `parellel_apply` does.

I also benchmarked swifter pandarallel and even looked at dask (but didn't try it).

Once I settled with pandaparallel, it just worked, and i removed a ton of code.